torchqdynamics/utils.py

```diff
@@ -120,19 +136,124 @@ def trace(rho: Tensor) -> Tensor:
     return torch.einsum('...ii', rho)
 
 
-def expect(operator: Tensor, state: Tensor) -> Tensor:
-    """Compute the expectation value of an operator on a quantum state or density
-    matrix. The method is batchable over the state, but not over the operator.
+def ptrace(
+    x: Tensor, dims_kept: Union[int, Tuple[int, ...]], hilbert_shape: Tuple[int, ...]
+) -> Tensor:
+    """Compute the partial trace of a state vector or density matrix, keeping only
+    dimensions `dims_kept`. The Hilbert space structure should be specified with
+    `hilbert_shape`.
 
     Args:
-        operator: tensor of shape (n, n)
-        state: tensor of shape (..., n, n) or (..., n)
+        x: Tensor of size `(..., n, 1)` or `(..., n, n)`
+        dims_kept: Int or tuple of ints containing the dimensions to keep for the
+            partial trace.
+        hilbert_shape: Tuple of ints specifying the dimensions of each mode in the
+            Hilbert space tensor product.
+
+    Returns:
+        Tensor of size `(..., m, m)` with `m <= n` containing the partially traced out
+            state vector or density matrix.
+
+    Example:
+        >>> rho = tq.kron(tq.coherent_dm(20, 2.0), tq.fock_dm(2, 0), tq.fock_dm(3, 1))
```

Comment:
I agree that I would prefer `tensor` to `kron`, but I feared it would be misleading together with `torch.tensor`. `tq.tensor` and `torch.tensor` would have completely different behavior.

What about `tq.tensprod` ? Different than qutip but the naming is quite clear IMO.

Reply:
Right, I'm not too sure I guess `kron` is maybe the best then. A fun alternative would be `otimes`. 😋

Reply:
Let's discuss it today!